Feature download parkings

matttsss opened the pull request:
## What is this PR?
This PR aims to enable the download and management of parkings in zones for offline mode.
It adds two new functions to the `ParkingViewModel`: 
* `downloadZone(...)` that downloads the parkings for a certain zone
* `deleteZone(...)`  that deletes the downloaded parkings that are within that zone

### Main changes:
* `Parking` now has a new attribute: `tile` which identifies to what tile it belongs
* `Tile` is now a type alias for `String`, this simplifies the management and creation of tiles, the "static" functions that were associated to it are now in a new object `TileUtils`.
* A new interface `OfflineParkingRepository` that extends `ParkingRepository`, it adds the API for dowloading / deleting zones
* `ParkingViewModel` has a new attribute of type `OfflineParkingRepository`

### Testing the database:
The Unit/Android tests cover the database by testing the `OfflineParkingRepositoryRoom`.
When the app is running, it is possible to inspect the database via the "App Inspector" as explained here:
https://developer.android.com/studio/inspect/database

## Code coverage:
- `com.github.se.cyrcle.model.parking.offline`: 100%
- `ParkingViewModel`: 86%
- Global:
![image](https://github.com/user-attachments/assets/d63cd252-04d6-433a-8d99-90c316c96b43)

andrepeiry commented on app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt:
```diff
@@ -0,0 +1,230 @@
+
+  @After
+  fun tearDown() {
+    database.close()
```
*[TEMPORARY COMMENT WHILE ANDROID STUDIO IS SYNCING]*
Wouldn't it be cleaner to also destroy the database as well ? I feel like this can be some bug source if we don't delete the database between tests. I'm ready to believe that JUnit recreates the whole environment each time, but I'm not an expert on this subject.

matttsss replied:
Mhhhh I'm not to sure. I think that because the `@Before` creates a new database, it ovewrites the database already created and held in that reference. Since those test databases are in memory, they will simply be deleted by the garbage collector?

LouisJamesVasseur replied:
Would you be willing to test both and report back? This could be a MAJOR error source so I'd be grateful if you tried before merging the PR. I am however not an expert either, so if you're confident I'll take your word for it

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:
For this file, I'm bothered with the following :
- Documentation is missing on most of the function
- Exception handling could be done at this level, as one of the two following, notably for the enum, in which the ordinal can be corrupt and bigger than the number of entries
  - change the function to the following scheme `deserializeOrDefault(ordinal, default) { try {deserialize} catch(E) {return default} }`
  - Throw a "proper" deserialization error instead of relying on the "IndexOutOfBound` when dealing with enums.

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -194,31 +189,25 @@ class ParkingViewModel(
         updateClosestParkings(--nbRequestLeft)
 
-        // Get parkings from repository
+        // Cache miss
```
Just a quick note (non-blocking), is that an explanation for an `if-else` statement should be written as follow
```
if (condition) {
    // Explanation
    ... code ...
} else {
    // Explanation
    ... code ...
}
```
Instead of above the branching, as this would better indicate the whole thing
```
// Absolute value of x
if (x < 0) {
    x =-x;
} else {
    x = x;
}
```

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -659,18 +648,61 @@ class ParkingViewModel(
+                      zone.boundingBox.southwest(), zone.boundingBox.northeast())
+              else emptySet()
+            }
+            .toSet()
```
I haven't tried myself, but could this have been obtain more easily with a `.filter` ?

matttsss replied:
I guess? But since there is only one element to filter out, I'm not sure if it would be that useful...

LouisJamesVasseur replied:
I think the flatMap as is is appropriate

andrepeiry replied:
Well then, let this stay as is, it was merely a suggestion ^^

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/model/parking/Tile.kt:
```diff
@@ -1,128 +1,139 @@
+   * @param tile the tile to get the bottom left point from
+   * @return the bottom left point of the tile
+   */
+  fun getBottomLeftPoint(tile: Tile): Point? {
```
The documentation for `getBottomLeftPoint` as well as `getTopRightPoint` could be improve such as
```
@param tile a valid tile to get the bottom left point from.
@return the bottom left point of the tile, null if tile is not valid.
```

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepositoryFirestore.kt:
```diff
@@ -68,12 +71,19 @@ class ParkingRepositoryFirestore @Inject constructor(private val db: FirebaseFir
       onFailure: (Exception) -> Unit
   ) {
+    // Get the bottom left and top right points of the tile
+    // This is a temporary solution for the time that `Parking.tile` gets added on Firestore
```
Please mark this with a `TODO` or another easy searchable string so that we can find them later ^^

andrepeiry commented on app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepository.kt:
Thank you for adding the documentation on `getReportsForParking` and `addReport` <3

andrepeiry reviewed (changes requested):
# Good code in general

Your code is readable, and even if the PR is quite hefty, it is not a pain to read.

## Some improvements

I'm a bit bothered by the `Adaptater`s

- Documentation is missing on most of the function
- Exception handling could be done at this level, as one of the two following, notably for the enum, in which the ordinal can be corrupt and bigger than the number of entries (also don't forget to add some tests on this)
   - change the function to the following scheme `deserializeOrDefault(ordinal, default) { try {deserialize} catch(E) {return default} }`
   - Throw a "proper" deserialization error instead of relying on the `IndexOutOfBound` when dealing with enums.

Also see some inline comments I left inside the code

LouisJamesVasseur commented on app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt:
```diff
@@ -9,31 +9,52 @@ import androidx.compose.ui.test.junit4.createComposeRule
 import org.mockito.Mockito.mock
 import org.mockito.kotlin.verify
 
+@RunWith(AndroidJUnit4::class)
```
any idea why it wasn't using the runWith before?

matttsss replied:
No, but it can only help to have them

LouisJamesVasseur replied:
Agree with you there

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/io/serializer/LocationAdapter.kt:
As @andrepeiry mentions, documentation is partly missing here

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:
```diff
@@ -12,37 +17,50 @@ import com.google.gson.reflect.TypeToken
+  }
+
+  @TypeConverter
+  fun deserializeParkingCapacity(ordinal: Int): ParkingCapacity {
```
Yeah I definitely agree w/ @andrepeiry here. If we're dealing w/ indices in a table we need to error handle... it'd be quite a shame for the app to consistenly crash if we make a change at this level

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:
```diff
@@ -12,37 +17,50 @@ import com.google.gson.reflect.TypeToken
+  }
+
+  @TypeConverter
+  fun deserializeParkingRackType(ordinal: Int): ParkingRackType {
```
same comment as above

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:
```diff
@@ -12,37 +17,50 @@ import com.google.gson.reflect.TypeToken
+
+  @TypeConverter
+  fun deserializeParkingProtection(ordinal: Int): ParkingProtection {
+    return ParkingProtection.entries[ordinal]
```
here too :)

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/di/DatabaseModule.kt:
```diff
@@ -0,0 +1,30 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+object DatabaseModule {
```
Maybe some doc here?

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -194,31 +189,25 @@ class ParkingViewModel(
-        parkingRepository.getParkingsBetween(
-            tile.bottomLeft,
-            tile.topRight,
+        tilesToParking[tile] = emptyList()
```
Why is this here? This is "false" branching of the if condition so that means that tilesToParking[tile] is already an empty list, no?

matttsss replied:
Since requests to Firestore are asynchronous, if we have a second call to `getParkingsInRect`, it will make a second request to firestore if the first one did not complete before. This way the empty list marks this tile as "already" queried

LouisJamesVasseur replied:
Good implementation then!

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:
```diff
@@ -659,18 +648,61 @@ class ParkingViewModel(
+    val tilesToDelete =
+        TileUtils.getAllTilesInRectangle(
+            zoneToDelete.boundingBox.southwest(), zoneToDelete.boundingBox.northeast())
+    offlineParkingRepository.deleteTiles(tilesToDelete - tilesToKeep) {
```
Again, a quick error handle to check that tilesToDelete > tilesToKeep would be nice

matttsss replied:
Since those are sets, the difference can only be the empty set in the worst case :)

LouisJamesVasseur replied:
Good point!

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt:
documentation here please :)

matttsss replied:
It's inherited from the interfaces :)

LouisJamesVasseur replied:
Yeah okay, I see. No need then

LouisJamesVasseur commented on app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneSelectionScreen.kt:
```diff
@@ -96,19 +100,30 @@ fun ZoneSelectionScreen(
+                {
+                  // On failure, avoid keeping stale zone
+                  Zone.deleteZone(zone, context)
+                  Toast.makeText(context, "Error downloading zone", Toast.LENGTH_SHORT).show()
```
Should be in strings.xml

matttsss commented:
@andrepeiry @LouisJamesVasseur I hear your points on the serialization for the parking enums. However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you?
Also, I did not bother with documenting those `@TypeConverter` functions since it is quite straight forward what they do, but I could add that if necessary.

LouisJamesVasseur reviewed (commented):
VERY good work overall. The length of the PR definitely makes sense and I think you did well to put all of this into 1 PR. Some light implementation comments on my end but I think @andrepeiry said most of what I wanted to say already so I'll leave it at that. Mind the comments and I will approve!

LouisJamesVasseur commented:
> @andrepeiry @LouisJamesVasseur I hear your points on the serialization for the parking enums. However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you? Also, I did not bother with documenting those `@TypeConverter` functions since it is quite straight forward what they do, but I could add that if necessary.

The arbitrary enum solution works for me, I think what you're suggesting makes sense. Understood on the documentation, honestly it was more out of concern about the documentation aspect of M3 since you added a lot of code here. I'll leave it up to your opinion if you want to add the doc you can otherwise it's non-blocking on my end

andrepeiry commented:
> However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you?

It's bothering me, as such exception could crash the app (I know it won't technically happens, as data corruption should be a quite negligible risk). Indeed, `if (corruption) return default` seems like a good enough solution for me ^^
If they are only used internally, maybe mark them as `private` ?

> Also, I did not bother with documenting those @TypeConverter functions since it is quite straight forward what they do, but I could add that if necessary.

I know documentation is as unloved as unit tests (for most human at least 😉), but it is always appreciated, and if I remember correctly, you are using GitHub Copilot, which is a great tool to autocomplete minimal documentation

matttsss commented:
Thank you both,  I have addressed your comments if you want to check.
@andrepeiry Concerning marking the `@TypeConverters` as `private`, that will not be possible since `Room` needs to access them from outside.

LouisJamesVasseur reviewed (approved):
Thanks for addressing the changes I requested! Great work!

andrepeiry reviewed (approved):
Thank you for addressing our comments, this PR is good to go !
